### PR TITLE
Szz unleashed blamer

### PR DESCRIPTION
**Pull request by Chrestopher**
SZZ blamer using temporary directories

used for one commit.

### CODE REVIEW

File: archeogitvsszz/utilities.py [hunk 1]
```diff
@@ -10,3 +10,33 @@ class YAML:
...
+
+class Calculation:
+    @staticmethod
+    def get_recall_and_precision(contributors, ground_truth):
+        true_positives = len(contributors & ground_truth)
+
+        false_negatives = 0
+        for contributor in ground_truth:
+            if contributor not in contributors:
+                false_negatives += 1
```
Comment (nuthanmunaiah) [inline archeogitvsszz/utilities.py:23]:
```suggestion
        false_negatives = len(ground_truth - contributors)
```

False negative is a contributing commit which was not predicted by a blamer (i.e. it was falsely marked as negative). In other words, it is all commits in `ground_truth` that are _not in_ `contributors`.

File: archeogitvsszz/utilities.py [hunk 2]
```diff
@@ -10,3 +10,33 @@ class YAML:
...
+
+        false_positives = 0
+        for contributor in contributors:
+            if contributor not in ground_truth:
+                false_positives += 1
```
Comment (nuthanmunaiah) [inline archeogitvsszz/utilities.py:28]:
```suggestion
        false_positives = len(contributors - ground_truth)
```

False positive is a contributing commit which was incorrectly predicted by a blamer (i.e. it was falsely marked as positive). In other words, it is all commits in `contributors` that are _not in_ `ground_truth`.

File: archeogitvsszz/utilities.py [hunk 3]
```diff
@@ -10,3 +10,33 @@ class YAML:
...
+
+        if true_positives + false_negatives != 0:
+            recall = true_positives / float(true_positives + false_negatives)
+        else:
+            recall = 0.0
```
Comment (nuthanmunaiah) [inline archeogitvsszz/utilities.py:33]:
```suggestion
        recall = 0.0
        if true_positives + false_negatives != 0:
            recall = true_positives / float(true_positives + false_negatives)            
```

File: archeogitvsszz/utilities.py [hunk 4]
```diff
@@ -10,3 +10,33 @@ class YAML:
...
+
+        if true_positives + false_positives != 0:
+            precision = true_positives / float(true_positives + false_positives)
+        else:
+            precision = 0.0
```
Comment (nuthanmunaiah) [inline archeogitvsszz/utilities.py:38]:
```suggestion
        precision = 0.0
        if true_positives + false_positives != 0:
            precision = true_positives / float(true_positives + false_positives)
```

File: archeogitvsszz/vulnerabilities.py [hunk 5]
```diff
@@ -1,14 +1,49 @@
...
+        self._cves = self.get_all_file_names()
+        self._index = 0
 
     def __iter__(self):
-        pass
+        return iter(self._cves)
 
     def __next__(self):
-        pass
+        self._index += 1
```
Comment (nuthanmunaiah) [inline archeogitvsszz/vulnerabilities.py:19]: Wouldn't incrementing `self._index` before getting the current value at `self._index` miss returning the first element?

File: main.py [hunk 6]
```diff
@@ -12,7 +13,7 @@ def _configure_logging(configuration):
     dictConfig(configuration['logging'])
 
 
-def main():
+if __name__ == '__main__':
```
Comment (nuthanmunaiah) [inline main.py:16]: I know you mentioned that having a `main()` and calling it from `if __name__ == '__main__'` did not work. Why was that again?

File: main.py [hunk 7]
```diff
@@ -22,6 +23,7 @@ def main():
 
     repository = Repository(arguments.repository)
     vulnerabilities = Vulnerabilities(arguments.vulnerabilities)
+    all_files = vulnerabilities.get_all_file_names()
```
Comment (nuthanmunaiah) [inline main.py:26]: Is this here by mistake? Why would we need the list of all files anywhere outside of the analyzer?

**Review by nuthanmunaiah — CHANGES_REQUESTED** [review]
Could you add a `pytest` test for the code that computes precision and recall?

File: main.py (hunk 6 above)
Comment (Chrestopher) [reply main.py:16]: It had something to do with multiprocessing, but I just moved multiprocessing into analyze so we can revert back to having `main()`

File: main.py (hunk 7 above)
Comment (Chrestopher) [reply main.py:26]: Yeah, I had it here before, but I forgot to move it when I moved everything into analyzer.

File: archeogitvsszz/vulnerabilities.py (hunk 5 above)
Comment (Chrestopher) [reply archeogitvsszz/vulnerabilities.py:19]: Hmm, this is my first time using `__next__`, but it appears that all of the cves were processed.

**Pull request merged**